frr-tools/status/controller/bgpsessionstate_controller.go

```diff
@@ -0,0 +1,250 @@
+// SPDX-License-Identifier:Apache-2.0
+
+package controller
+
+import (
+	"context"
+	"net/netip"
+	"reflect"
+	"strings"
+	"time"
+
+	frrk8sv1beta1 "github.com/metallb/frr-k8s/api/v1beta1"
+	"github.com/metallb/frr-k8s/internal/frr"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+)
+
+const (
+	nodeLabel       = "frrk8s.metallb.io/node"
+	peerLabel       = "frrk8s.metallb.io/peer"
+	vrfLabel        = "frrk8s.metallb.io/vrf"
+	noBFDConfigured = "N/A"
+)
+
+type BGPPeersFetcher func() (map[string][]*frr.Neighbor, error)
+
+// BGPSessionStateReconciler reconciles a BGPSessionState object.
+type BGPSessionStateReconciler struct {
+	client.Client
+	BGPPeersFetcher
+	Logger       log.Logger
+	NodeName     string
+	Namespace    string
+	DaemonPod    *corev1.Pod
+	ResyncPeriod time.Duration
+}
+
+// +kubebuilder:rbac:groups=frrk8s.metallb.io,resources=bgpsessionstates,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=frrk8s.metallb.io,resources=bgpsessionstates/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=frrk8s.metallb.io,resources=frrnodestates,verbs=get;list;watch
+// +kubebuilder:rbac:groups=frrk8s.metallb.io,resources=frrnodestates/status,verbs=get
+// +kubebuilder:rbac:groups="",resources=nodes,verbs=get;list;watch
+
+func (r *BGPSessionStateReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	level.Info(r.Logger).Log("controller", "BGPSessionState", "start reconcile", req.NamespacedName.String())
+	defer level.Info(r.Logger).Log("controller", "BGPSessionState", "end reconcile", req.NamespacedName.String())
+
+	l := frrk8sv1beta1.BGPSessionStateList{}
+	err := r.Client.List(ctx, &l, client.MatchingLabels{nodeLabel: r.NodeName})
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	existing := map[string]map[string]*frrk8sv1beta1.BGPSessionState{} // vrf -> peer -> existing status
+	for _, s := range l.Items {
```

Comment:
let's wrap this in a function (even returning duplicates to delete or looping twice)

Reply:
done